### PR TITLE
Search in Doc Explorer

### PR DESCRIPTION
**Pull request by asiandrummer**
Search functionality in Documentation Explorer.
It's pretty basic with searching/getting results/clicking links - I'll incrementally add in more details in upcoming PRs.

EDIT: added no-search-result case and minimum character length

### CODE REVIEW

File: src/components/DocExplorer.js [hunk 1]
```diff
@@ -45,8 +45,8 @@ export class DocExplorer extends React.Component {
   // Public API
   showDoc(typeOrField) {
     var navStack = this.state.navStack;
-    var isCurrentlyShown =
-      navStack.length > 0 && navStack[navStack.length - 1] === typeOrField;
+    var isCurrentlyShown = navStack.length > 0 &&
```
Comment (asiandrummer) [inline src/components/DocExplorer.js]: This is an unnecessary change - reverting.

**Comment by asiandrummer** [comment]
screen shot 2016-01-15 at 4 17 58 pm
screen shot 2016-01-15 at 4 17 50 pm

**Comment by asiandrummer** [comment]
screen shot 2016-01-15 at 4 43 39 pm
screen shot 2016-01-15 at 4 39 30 pm

File: src/components/DocExplorer.js [hunk 2]
```diff
@@ -56,7 +56,10 @@ export class DocExplorer extends React.Component {
   constructor() {
     super();
 
-    this.state = { navStack: [] };
+    this.state = {
+      navStack: [],
+      searchValue: null
```
Comment (leebyron) [inline src/components/DocExplorer.js]: I think searchValue should probably be part of the nav stack rather than it's own thing, so that you can search, click on something, click back and be where you were before.

This could be done in a separate PR though.

File: src/components/DocExplorer.js [hunk 3]
```diff
@@ -103,12 +114,25 @@ export class DocExplorer extends React.Component {
...
+      title = 'Search Results';
+      content = this.state.searchValue.length > 2 ?
+        <SearchDoc
+          searchValue={this.state.searchValue}
+          schema={schema}
+          onClickType={this._onClickTypeOrField}
+          onClickField={this._onClickTypeOrField}
+        /> :
+        <span className="doc-alert-text">
+          Minimum search length is 3 characters.
```
Comment (leebyron) [inline src/components/DocExplorer.js]: I don't think we need to explicitly show this on the screen. Also, in a separate PR later we may want to remove this restriction so you could search for things like "id" but perhaps just change the search algorithm so it's not overwhelmed by < 3 ch searches.

File: src/components/DocExplorer.js [hunk 4]
```diff
@@ -141,6 +165,15 @@ export class DocExplorer extends React.Component {
           </div>
         </div>
         <div className="doc-explorer-contents">
+          {!typeOrField &&
+            <label className="search-box-outer">
+              <input className="search-box-input"
+                onKeyUp={event => this._onSearch(event)}
+                type="text"
+                name="SearchBox"
```
Comment (leebyron) [inline src/components/DocExplorer.js]: I don't think you need `name=` here

File: src/components/DocExplorer.js [hunk 5]
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
...
+  render() {
+    const searchValue = this.props.searchValue;
+    const schema = this.props.schema;
+    const onClickType = this.props.onClickType;
+    const onClickField = this.props.onClickField;
+
+    const typeMap = schema.getTypeMap();
+
+    // Compose a search index for all possible arguments from every types.
+    let possibleArguments = [];
```
Comment (leebyron) [inline src/components/DocExplorer.js]: This isn't used anywhere below, maybe we should remove it for this PR.

Also, we don't want to compute this index on every render since it's likely to be expensive for larger schema. We only want to compute an index once per schema.

File: src/components/DocExplorer.js [hunk 6]
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
...
+    let matchedArgs = [];
+
+    Object.keys(typeMap).forEach(typeName => {
+      const type = typeMap[typeName];
+      let matchedOn = [];
+      if (typeName.indexOf(searchValue) !== -1) {
+        matchedOn.push('Type Name');
+      }
+
+      if (type.description && type.description.indexOf(searchValue) !== -1) {
```
Comment (leebyron) [inline src/components/DocExplorer.js]: I would remove this for this initial PR - I'm concerned that searching over descriptions will be expensive for larger schema, and perhaps we'll want to do a specialized index for this. Let's come back to it in a separate PR once the main search functionality is working.

File: src/components/DocExplorer.js [hunk 7]
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
...
+                  onClick={event => onClickField(field, type, event)}>
+                  {field.name}
+                </a>
+                {' on '}
+                <TypeLink type={type} onClick={onClickType} />
+              </div>
+            );
+          }
+
+          if (field.args && field.args.length) {
```
Comment (leebyron) [inline src/components/DocExplorer.js]: probably want `else if` here so the same field doesnt appear twice if an argument also matches

File: src/components/DocExplorer.js [hunk 8]
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
...
+                {' on '}
+                <TypeLink type={type} onClick={onClickType} />
+              </div>
+            ));
+          }
+        });
+      }
+    });
+
+    if (matchedTypes.length === 0 &&
```
Comment (leebyron) [inline src/components/DocExplorer.js]: style nit:

```js
if (matchedTypes.length === 0 &&
    matchedFields.length === 0 &&
    matchedArgs.length === 0) {
  return ...
}

File: src/components/DocExplorer.js [hunk 9]
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
...
+                <TypeLink type={type} onClick={onClickType} />
+              </div>
+            );
+          }
+
+          if (field.args && field.args.length) {
+            let matches = field.args.filter(arg => {
+              return arg.name.indexOf(searchValue) !== -1;
+            });
+            Array.prototype.push.apply(matchedArgs, matches.map(arg =>
```
Comment (leebyron) [inline src/components/DocExplorer.js]: Since you're rendering fields here, you might want to add them to `matchedFields`. Also, rather than rendering the same field multiple times for each matched arg, you should just render all the matched args within this one field, moving the `matches.map` inside the `( )` selection below.

**Comment by leebyron** [comment]
This is awesome! There are some comments inline, but this is pretty close to landing as an initial version of search.

File: src/components/DocExplorer.js (hunk 2 above)
Comment (asiandrummer) [inline src/components/DocExplorer.js]: Exactly; I would like the search action to be a part of navStack - something like:
navStack = [
  testType: { ... },
  graphiql_search: { value: ... },
];

**Comment by asiandrummer** [comment]
Responded to comments:
- removed ch restriction for now
- got rid of the argument indexing, turns out we didn't need it after all
- removed specific match types and included wholesome within `search results`
- a field with multiple argument matches

**Comment by asiandrummer** [comment]
screen shot 2016-01-21 at 2 56 40 pm

File: src/components/DocExplorer.js (hunk 2 above)
Comment (leebyron) [inline src/components/DocExplorer.js]: A related note about navStack for the future: it would be cool if scroll position were also maintained as part of the back stack. Today if you click from a long scrolling view to another long scrolling view your scroll position is maintained when it should probably be reset to "0", but then when going "back" the original scroll position should be restored. It's a fairly small detail, but it will help make navigating larger schema easier on the mind.

File: src/components/DocExplorer.js [hunk 10]
```diff
@@ -103,12 +114,21 @@ export class DocExplorer extends React.Component {
           field={typeOrField}
           onClickType={this._onClickTypeOrField}
         />;
+    } else if (this.state.searchValue) {
+      title = 'Search Results';
+      content = <SearchDoc
```
Comment (leebyron) [inline src/components/DocExplorer.js]: nit: use the same styling here as in the else block below, line break and indent.

File: src/components/DocExplorer.js [hunk 11]
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
   }
 }
 
+// Render Search Results
+class SearchDoc extends React.Component {
+  shouldComponentUpdate(nextProps) {
+    return (this.props.schema !== nextProps.schema) ||
+      (this.props.searchValue !== nextProps.searchValue);
```
Comment (leebyron) [inline src/components/DocExplorer.js]: nit: no need for the wrapping () on each of these terms.

File: src/components/DocExplorer.js [hunk 12]
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
...
+    const searchValue = this.props.searchValue;
+    const schema = this.props.schema;
+    const onClickType = this.props.onClickType;
+    const onClickField = this.props.onClickField;
+
+    const typeMap = schema.getTypeMap();
+
+    let matchedTypes = [];
+    let matchedFields = [];
+    let matchedArgs = [];
```
Comment (leebyron) [inline src/components/DocExplorer.js]: this is no longer being used, rm

File: src/components/DocExplorer.js [hunk 13]
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
...
+              <div className="doc-category-item">
+                <a className="field-name"
+                  onClick={event => onClickField(field, type, event)}>
+                  {field.name}
+                </a>
+                {'('}
+                <span>
+                  {matches.map(arg =>
+                    <span className="arg" key={arg.name}>
+                        <span className="arg-name">{arg.name}</span>
```
Comment (leebyron) [inline src/components/DocExplorer.js]: nit: indented 2sp too far

File: src/components/DocExplorer.js [hunk 14]
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
...
+                  onClick={event => onClickField(field, type, event)}>
+                  {field.name}
+                </a>
+                {' on '}
+                <TypeLink type={type} onClick={onClickType} />
+              </div>
+            );
+          } else if (field.args && field.args.length) {
+            let matches = field.args.filter(arg => {
+              return arg.name.indexOf(searchValue) !== -1;
```
Comment (leebyron) [inline src/components/DocExplorer.js]: you can make this a directly returned expression instead of a function block

```
const matches = field.args.filter(arg => arg.name.indexOf(searchValue) !== -1);
```

**Comment by leebyron** [comment]
This looks great. I have a few other comments in there. One overarching minor thing: it's nice to use `const` instead of `let` as the default for defining variables, and use `let` as the exception when you need to reassign it in the future.

Once you've covered all the comments and updated it, merge away!

File: src/components/DocExplorer.js [hunk 15]
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
...
+            });
+            matchedFields.push(
```
Comment (asiandrummer) [inline src/components/DocExplorer.js]: This will allow search results to match arguments/render fields every time. Fixed:
```
if (matches.length > 0) {
```

**Comment by asiandrummer** [comment]
Thanks @leebyron!
Next diffs will include navStack integration and a newer format in Doc Explorer.
